README.md
Trillek Virtual Computer Lib
============================

Using these [specs](https://github.com/trillek-team/trillek-computer)

The design of the Virtual Computer allows to plug and use different CPUs. Actually is the TR3200.

This repo contains of the Trillek Virtual Computer lib and some extra tools, including a toy emulator build over the lib.

| GNU/Linux                                        | Windows (VStudio 2012)  |
|--------------------------------------------------|-------------------------|
| [![Build Status](https://travis-ci.org/trillek-team/trillek-vcomputer-module.svg?branch=develop)](https://travis-ci.org/trillek-team/trillek-vcomputer-module) | [![Build status](https://ci.appveyor.com/api/projects/status/420wp0vh8tnsuvjn/branch/develop?svg=true)](https://ci.appveyor.com/project/Zardoz89/trillek-vcomputer-module-r5m16/branch/develop) |

COMPILING
---------
We are using CMake and C++ 11 (Vs2013/14 , G++ >= 4.7, LLVM). So if you are in GNU/Linux you could compile with this :

    mkdir build
    cd build
    cmake ..
    make

By default it will generate a dynamic and static libraries and compile tools and tests.

If you not wish to compile the tools, then you should use `cmake -DBUILD_TOOLS_VCOMPUTER=False ..`. Also you can control if you like to have virtual screen and audio with BUILD_TOOLS_SCREEN and BUILD_TOOLS_AUDIO (By default both are true)

If you not wish to compile the tests, then you should use `cmake -DBUILD_TESTS_VCOMPUTER=False ..`

Plus you can control if you wish release or debug building using `-DCMAKE_BUILD_TYPE=Release` or `-DCMAKE_BUILD_TYPE=Debug` flag in cmake (note that if you uses VStudio, it would change it acording to your solution output mode)

In windows, you should add the "-g" parameter with the apropiated generator for VS2013 (you can get the list running cmake --help), or you can use the CMake GUI. With this, you can generate VS2013/4 project/solution files that you can open and compile, or use msbuild from a develop console. On window, to get wroking video and sound, you should put all necesary library and include files on a "lib" directory on the root dir of the source code. You could download a zip file with all necesary library and includes files for Windows, from http://cpu.zardoz.es/trillek-win32-lib.zip . If you use this file, you need to set the flag USE_STATIC_GLEW to true.

Rememeber to copy/link the assets folder to the directory build/bin/ (on Linux/OSX) or build/bin/Debug/ or build/bin/Release/ 
INSTALLING
----------

You could do a ```make install``` (or run Install solution on Visual Studio) to install the toy emulator + tools + library.

WHAT IT DOES ACTUALLY
------------------
### Tools

#### vm

Is a program that uses the Trillek Virtual Computer lib to run a TR3200 emulation. Can run in step mode or in "run mode" where it executes all the program without stopping. Needs C++ 11 *std::chrono compilance* for measuring times; **GLFW3 + GLM + OpenGL 3.3** is needed if you desire to see the virtual screen.

- Can load a little endian binary file with a ROM image with a max size of 32KiB.
- Step mode, has an on-line disassembler. Each time that you press enter or 's' or 'S'
  keys, one instruction is executed, and the status of registers and stack, is
  shown. '**q**' ends the execution. '**r**' switch to Run mode
- Run mode, It tries to enforce CPU Clock speed. Only stops by doing '**Ctrl+C**' or closing the virtual screen window, or when finds a breakpoint.

It accepts this parameters:
```
Virtual Computer toy Emulator

Usage:
        ./vm -r romfile [other parameters]

Parameters:
        -r file or --rom file : RAW binary file for the ROM (32 KiB Max)
        -d file or --disk file : Disk file
        --nvram file : Non volatile RAM file
        -c val or --cpu val : Sets the CPU to use, from "tr3200" or "dcpu-16n"
        -x or --exec : Run the computer in normal mode without asking
        -m val or --mem val : How much RAM the computer will have, in KiB. Must be between 32 and 1024 and will be rounded to a multiple of 32
        --time : Show timing and speed info while running.
        --clock val : CPU clock speed in Khz. Must be 100, 250, 500 or 1000.
        -b val : Inserts a breakpoint at address val (could be hexadecimal or decimal).
        --ext-keys : Allow extra (non-standard) keycodes with virtual keyboard.
        -h or --help : Shows this help
```

For example, if you run vc on the directory were is generated when you compile the project, you could simply :
```./vc -r ../../asm/hello.ffi -x```
And it would run the "Hello world" program.

#### pbm2font

Is a small tool that can generate a hexadecimal representation of a user font for the TDA display card. Uses as source image, a b&w ASCII **pbm** file (Gimp can generate it). It expects that the image have a width and height multiple of 8, as it divides the image in cells of 8x8 pixels each being a glyph of the font.

### makedisk

Is a tool that can generate floppy disk images of different sizes and could write on a disk image, raw binary data from a file.

```
makedisk

Usage:
    makedisk -o outputfile [other parameters]

Parameters:
    -o file : Output file
    -i file : RAW binary file to be included on the floppy disk image
    -s sectors : Sectors per track (8 to 36)
    -t tracks : Tracks per side (between 40 and 80)
    -1 : Creates a floppy with only a sigle side
    -160 : Creates a 160KiB disk with a single side, 8 sectors and 40 tracks.
    -320 : Creates a 320KiB disk with 8 sectors and 40 tracks. Default floppy size
    -360 : Creates a 360KiB disk with 9 sectors and 40 tracks.
    -640 : Creates a 640KiB disk with 8 sectors and 80 tracks.
    -720 : Creates a 720KiB disk with 9 sectors and 80 tracks.
    -1200 : Creates a 1200KiB disk with 15 sectors and 80 tracks.
    -h : Shows this help
```

### Tests

#### benchmark

Is a program that runs a quick and dirty benchmark to measure the performance of the Trillek Virtual Computer lib. Needs C++ 11 *std::chrono compilance* for measuring times.

- Can load a list of little endian binary files as ROMs for each Virtual Computer.
- Aceepts a final optional parameter with the number of CPUs to run, if not especified, then will be 1000.
- Each Virtual Computer have the same set of devices plugged and the same amount of RAM
- Executes a random number of cycles before begin the benchmark to make more realistic.
- Runs CPUs at different clock speeds, aroung 70% are at 100KHz, 20% at 200KHz, 5% at 500KHz and 1% at 1Mhz.
- Tries to execute Virtual Computers with the TR3200 CPUs in a single thread at what your computer can give. It does not enforce a particular speed, but it gives information of the emulation speed. To know how many Virtual Computer you can run in your computer, tune the number of CPUs to get a speed around 110%, and these value will be the number of CPUs adecuate for your machine.
- Only stops by doing 'Ctrl+C'.
- Try it with with **Release** compile mode, there is an apreciable difference.

#### unit test

If before calling CMake, you setup a enviroment variable called "GTEST_ROOT" with the path to GTest suite, then the unit tests could be compiled and will generate an executable file "unit_test". This allow to check if some basic code gets broken if someone try to edit the source code. Are far of being a exaustive tests, but it's far better that nothing.

To set a enviroment variable on GNU/Linux (and Mac): ```export GTEST_ROOT=...```

If you are using fish shell : ```set -x GTEST_ROOT /usr/src/gtest/```

In windows, you should do that in Control Panel, or if you run CMake on a PowerShell console, with : ```$env:GTEST_ROOT = $PWD.Path + "\gtest-1.7.0\"```

HOW I CAN CREATE PROGRAMS ?
---------------------------
Actually you have a few options :

- [Meisaka's WaveAsm](https://github.com/Meisaka/WaveAsm) to generate TR3200 machine code.
- [VASM](https://github.com/Zardoz89/VASM) is a portable and retargetable assembler. It have a TR3200 module that can generate flat binary files and VOBJ object files. It have support for assembly macros and other fancy stuff.
- [SmallerC](https://github.com/alexfru/SmallerC) is a small C compiler, that can generate TR3200 assembly code for VASM. Actually we not have yet a C Standard lib. Any help with it would be very apreciated.

ADJUNCT TR3200 ASM PROGRAMS
--------------------------
There are some TR3200 ASM programs, source code and binary files, in /asm directory. Specifically :

- type1.asm : Basic typing program (uses busy wait)
- test.asm : Some tests of TR3200 CPU compliance and checks. Should show an OK on screen.
- random.asm : Uses the RNG generator. Only to check that it works on step-mode of "vm"
- hwn.asm : List how many devices are pluged and what are.
- hello.asm : Hello world
- floppy.asm : Writes data to a floppy. You could check that "disk.dsk" file get filled with some data.
- clock.asm : Basic example of system clock using Timer0
- beep.asm : Generates some sound with the beeper.

Also, you can download an functional firmware from :

- [Zardoz89/trillek-firmware](https://github.com/Zardoz89/trillek-firmware) : An firmware that try to boot from an floppy and if it fails, executes a simple machine code monitor (an clone of Wozniak monitor of Apple 1 and Apple ][)
- [meisys](http://eciv.net/meisaka/meisys.ffi) : An firmware that try to boot from a floppy and shows some useful information. Also, sets a system clock. Actually only have the list command to list devices attached to the machine.


IMPLEMENTED DEVICES
-------------------

- TDA graphics card
- Generic Wester/Latin keyboard
- Beeper, RTC, RNG and Timer embed devices
- Floppy drive
- Dummy device (Used only to test stuff)
- Debug Serial Console (only to be used in develop of the virtual computer and Trillek itself)

TODO
----
See issues on github or [Trello board](https://trello.com/b/ONRIy13Q)


CHANGELOG.md
# Change Log

## [Unreleased](https://github.com/trillek-team/trillek-vcomputer-module/tree/HEAD)

[Full Changelog](https://github.com/trillek-team/trillek-vcomputer-module/compare/v0.5.1...HEAD)

**Fixed bugs:**

- Tools with OpenGL on OSX [\#58](https://github.com/trillek-team/trillek-vcomputer-module/issues/58)

**Closed issues:**

- TR3200 cycles were not updated with v0.4.0 and v0.4.1 specifications [\#54](https://github.com/trillek-team/trillek-vcomputer-module/issues/54)

**Merged pull requests:**

- Fix OSX and extra stuff [\#59](https://github.com/trillek-team/trillek-vcomputer-module/pull/59) ([Zardoz89](https://github.com/Zardoz89))

## [v0.5.1](https://github.com/trillek-team/trillek-vcomputer-module/tree/v0.5.1) (2015-02-19)

[Full Changelog](https://github.com/trillek-team/trillek-vcomputer-module/compare/v0.5.0...v0.5.1)

**Closed issues:**

- Skipping a 8-byte instruction should take 1 extra-cycle [\#56](https://github.com/trillek-team/trillek-vcomputer-module/issues/56)

- Build issues on OS X [\#53](https://github.com/trillek-team/trillek-vcomputer-module/issues/53)

**Merged pull requests:**

- Fix issue \#53 [\#57](https://github.com/trillek-team/trillek-vcomputer-module/pull/57) ([Zardoz89](https://github.com/Zardoz89))

- Many minor changes and package generation [\#55](https://github.com/trillek-team/trillek-vcomputer-module/pull/55) ([Zardoz89](https://github.com/Zardoz89))

## [v0.5.0](https://github.com/trillek-team/trillek-vcomputer-module/tree/v0.5.0) (2015-01-30)

[Full Changelog](https://github.com/trillek-team/trillek-vcomputer-module/compare/v0.4...v0.5.0)

**Fixed bugs:**

- Floppy drive device not matchs specs [\#43](https://github.com/trillek-team/trillek-vcomputer-module/issues/43)

**Closed issues:**

- Refactor DisassemblyTR3200 [\#42](https://github.com/trillek-team/trillek-vcomputer-module/issues/42)

- Drop DCPU16N [\#23](https://github.com/trillek-team/trillek-vcomputer-module/issues/23)

**Merged pull requests:**

- Fix appveyour [\#52](https://github.com/trillek-team/trillek-vcomputer-module/pull/52) ([Zardoz89](https://github.com/Zardoz89))

- :checkered\_flag: A lot of tweks for Windows [\#51](https://github.com/trillek-team/trillek-vcomputer-module/pull/51) ([Zardoz89](https://github.com/Zardoz89))

- Clean CMakeFile and add options [\#50](https://github.com/trillek-team/trillek-vcomputer-module/pull/50) ([Zardoz89](https://github.com/Zardoz89))

- NVRAM implementation [\#49](https://github.com/trillek-team/trillek-vcomputer-module/pull/49) ([Zardoz89](https://github.com/Zardoz89))

- Trap interface for CPUs -\>Fix Single Step on TR32 [\#48](https://github.com/trillek-team/trillek-vcomputer-module/pull/48) ([Zardoz89](https://github.com/Zardoz89))

- Fix fdd [\#47](https://github.com/trillek-team/trillek-vcomputer-module/pull/47) ([Zardoz89](https://github.com/Zardoz89))

- :lipstick: Refactor DisassemblyTR3200 [\#46](https://github.com/trillek-team/trillek-vcomputer-module/pull/46) ([Zardoz89](https://github.com/Zardoz89))

- Build static dynamic same time [\#45](https://github.com/trillek-team/trillek-vcomputer-module/pull/45) ([Zardoz89](https://github.com/Zardoz89))

- Updating develop branch [\#44](https://github.com/trillek-team/trillek-vcomputer-module/pull/44) ([Zardoz89](https://github.com/Zardoz89))

- Minor API changes [\#39](https://github.com/trillek-team/trillek-vcomputer-module/pull/39) ([Zardoz89](https://github.com/Zardoz89))

## [v0.4](https://github.com/trillek-team/trillek-vcomputer-module/tree/v0.4) (2014-12-25)

[Full Changelog](https://github.com/trillek-team/trillek-vcomputer-module/compare/v0.1.0...v0.4)

**Fixed bugs:**

- Chaining branchs not works correctly [\#35](https://github.com/trillek-team/trillek-vcomputer-module/issues/35)

**Closed issues:**

- OpenGL does not initiate??? [\#20](https://github.com/trillek-team/trillek-vcomputer-module/issues/20)

**Merged pull requests:**

- Engine fixups, DLL stuff, and TDA improvements [\#38](https://github.com/trillek-team/trillek-vcomputer-module/pull/38) ([Meisaka](https://github.com/Meisaka))

- Tda improvement [\#36](https://github.com/trillek-team/trillek-vcomputer-module/pull/36) ([Zardoz89](https://github.com/Zardoz89))

- TDA Hardware cursor [\#34](https://github.com/trillek-team/trillek-vcomputer-module/pull/34) ([Zardoz89](https://github.com/Zardoz89))

- Update to TR3200 v0.4.0 [\#33](https://github.com/trillek-team/trillek-vcomputer-module/pull/33) ([Zardoz89](https://github.com/Zardoz89))

- Fix overflow flag not cleared in ALU operations [\#32](https://github.com/trillek-team/trillek-vcomputer-module/pull/32) ([catageek](https://github.com/catageek))

- Fix IF instructions adding 1 more cycle when skipping [\#31](https://github.com/trillek-team/trillek-vcomputer-module/pull/31) ([catageek](https://github.com/catageek))

- Update cpu v0.3 [\#30](https://github.com/trillek-team/trillek-vcomputer-module/pull/30) ([Zardoz89](https://github.com/Zardoz89))

- Dynamic library stuff [\#37](https://github.com/trillek-team/trillek-vcomputer-module/pull/37) ([Zardoz89](https://github.com/Zardoz89))

## [v0.1.0](https://github.com/trillek-team/trillek-vcomputer-module/tree/v0.1.0) (2014-10-23)

[Full Changelog](https://github.com/trillek-team/trillek-vcomputer-module/compare/ms2.5...v0.1.0)

**Implemented enhancements:**

- Fix removing device / address handler. Improve address handlers container [\#4](https://github.com/trillek-team/trillek-vcomputer-module/issues/4)

**Fixed bugs:**

- Fix clock.asm or interrupts in improve-tr3200 branch, before merging to master [\#9](https://github.com/trillek-team/trillek-vcomputer-module/issues/9)

- Fix removing device / address handler. Improve address handlers container [\#4](https://github.com/trillek-team/trillek-vcomputer-module/issues/4)

**Closed issues:**

- cmake script should detect if gtest is already built [\#25](https://github.com/trillek-team/trillek-vcomputer-module/issues/25)


src/Aux.cpp
/**
 * Auxiliar functions
 * \file aux.cpp
 */

#include "Aux.hpp"
#include "VSFix.hpp"
#include "GKeyboard.hpp"


namespace vm {
  namespace aux {

  byte_t GLFWKeyToTR3200 (int key) {
    switch (key) {
      case -1:
        return vm::keyboard::KEY_UNKNOW;

      case 256:
        return vm::keyboard::KEY_ESC;
      case 257:
        return vm::keyboard::KEY_RETURN;
      case 258:
        return vm::keyboard::KEY_TAB;
      case 259:
        return vm::keyboard::KEY_BACKSPACE;
      case 260:
        return vm::keyboard::KEY_INSERT;
      case 261:
        return vm::keyboard::KEY_DELETE;

      case 262:
        return vm::keyboard::KEY_ARROW_RIGHT;
      case 263:
        return vm::keyboard::KEY_ARROW_LEFT;
      case 264:
        return vm::keyboard::KEY_ARROW_DOWN;
      case 265:
        return vm::keyboard::KEY_ARROW_UP;

      case 340:
      case 344:
        return vm::keyboard::KEY_SHIFT;
      case 341:
      case 345:
        return vm::keyboard::KEY_CONTROL;
      case 346:
        return vm::keyboard::KEY_ALT_GR;

      default:
        if (key <= 255) // Maps 1:1 GLFW keys
          return key;
        return vm::keyboard::KEY_UNKNOW;

    }
    return vm::keyboard::KEY_UNKNOW;
  }

  byte_t SDL2KeyToTR3200 (int key) {
    switch (key) {
      case 0:
        return vm::keyboard::KEY_UNKNOW;

      case 4:
      case 5:
      case 6:
      case 7:
      case 8:
      case 9:
      case 10:
      case 11:
      case 12:
      case 13:
      case 14:
      case 15:
      case 16:
      case 17:
      case 18:
      case 19:
      case 20:
      case 21:
      case 22:
      case 23:
      case 24:
      case 25:
      case 26:
      case 27:
      case 28:
      case 29:
        return (key - 4) + vm::keyboard::KEY_A;

      case 30:
      case 31:
      case 32:
      case 33:
      case 34:
      case 35:
      case 36:
      case 37:
      case 38:
        return (key - 30) + vm::keyboard::KEY_1;
      case 39:
        return vm::keyboard::KEY_0;

      case 40:
        return vm::keyboard::KEY_RETURN;
      case 41:
        return vm::keyboard::KEY_ESC;
      case 42:
        return vm::keyboard::KEY_BACKSPACE;
      case 43:
        return vm::keyboard::KEY_TAB;
      case 44:
        return vm::keyboard::KEY_SPACEBAR;
      case 45:
        return vm::keyboard::KEY_MINUS;
      case 46:
        return vm::keyboard::KEY_EQUAL;
      case 47:
        return vm::keyboard::KEY_LEFT_BRACKET;
      case 48:
        return vm::keyboard::KEY_RIGHT_BRACKET;
      case 49:
        return vm::keyboard::KEY_BACKSLASH;
      case 51:
        return vm::keyboard::KEY_SEMICOLON;
      case 52:
        return vm::keyboard::KEY_APOSTROPHE;
      case 53:
        return vm::keyboard::KEY_GRAVE_ACCENT;
      case 54:
        return vm::keyboard::KEY_COMA;
      case 55:
        return vm::keyboard::KEY_PERIOD;
      case 56:
        return vm::keyboard::KEY_SLASH;
      case 73:
        return vm::keyboard::KEY_INSERT;
      case 76:
        return vm::keyboard::KEY_DELETE;

      case 79:
        return vm::keyboard::KEY_ARROW_RIGHT;
      case 80:
        return vm::keyboard::KEY_ARROW_LEFT;
      case 81:
        return vm::keyboard::KEY_ARROW_DOWN;
      case 82:
        return vm::keyboard::KEY_ARROW_UP;

      case 224:
      case 228:
        return vm::keyboard::KEY_CONTROL;
      
      case 225:
      case 229:
        return vm::keyboard::KEY_SHIFT;
      
      case 230:
        return vm::keyboard::KEY_ALT_GR;

      default:
        return vm::keyboard::KEY_UNKNOW;
      
    }
    return vm::keyboard::KEY_UNKNOW;
  }


  } // end of namespace aux
} // end of namespace vm

